core/src/main/java/com/oath/oak/NativeAllocator/OakNativeMemoryAllocator.java

```diff
@@ -12,27 +12,36 @@
 import java.nio.ByteBuffer;
 import java.util.AbstractMap;
 import java.util.Comparator;
+
 import java.util.Map;
-import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.ConcurrentHashMap;
+
 import java.util.concurrent.ConcurrentSkipListSet;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
 import java.util.concurrent.atomic.AtomicLong;
 
 public class OakNativeMemoryAllocator implements OakMemoryAllocator {
 
     // When allocating n bytes and there are buffers in the free list, only free buffers of size <= n * RECLAIM_FACTOR will be recycled
     // This parameter may be tuned for performance vs off-heap memory utilization
     private static final int RECLAIM_FACTOR = 2;
+    public static final int INVALID_BLOCK_ID = 0;
 
-    // blocks allocated solely to this Allocator
-    private final ConcurrentLinkedQueue<Block> blocks = new ConcurrentLinkedQueue<>();
+    // mapping IDs to blocks allocated solely to this Allocator
+    private final ConcurrentHashMap<Integer,Block> blocks = new ConcurrentHashMap<>();
+    private final AtomicInteger idGenerator = new AtomicInteger(1);
 
     // free list of ByteBuffers which can be reused - sorted by buffer size, then by unique hash
-    private Comparator<Map.Entry<Long, ByteBuffer>> comparator =
-            Comparator.<Map.Entry<Long, ByteBuffer>, Integer>comparing(p -> p.getValue().remaining())
+
+    private Comparator<Map.Entry<Long, Slice>> comparator =
+            Comparator.<Map.Entry<Long, Slice>, Integer>comparing(p -> (p.getValue()).getByteBuffer().remaining())
             .thenComparing(Map.Entry::getKey);
-    private final ConcurrentSkipListSet<Map.Entry<Long,ByteBuffer>> freeList = new ConcurrentSkipListSet<>(comparator);
+    private final ConcurrentSkipListSet<Map.Entry<Long,Slice>> freeList = new ConcurrentSkipListSet<>(comparator);
+
     private final BlocksProvider blocksProvider;
     private Block currentBlock;
+    private int currentBlockID;
```

Comment:
why do we need currentBlockId ? can't we just read the content of idGenerator-1?